database/src/ops/output.rs

```diff
@@ -1,34 +1,213 @@
+        &env.open_db_ro::<NumOutputs>(tx_ro)?,
+        &env.open_db_ro::<RctOutputs>(tx_ro)?,
+    );
+
+    let mut vec = Vec::with_capacity(outputs.len());
+
+    for (amount, amount_index) in outputs {
+        let output = get_output_internal(
+            amount,
+            amount_index,
+            table_outputs,
+            table_key_images,
+            table_num_outputs,
+            table_rct_outputs,
+        )?;
+        vec.push(output);
+    }
+
+    Ok(vec)
+}
+
 /// TODO
-pub fn get_output() {
+#[inline]
+pub(super) fn get_output_internal(
+    amount: Amount,
+    amount_index: AmountIndex,
+    table_outputs: &(impl DatabaseRo<Outputs> + DatabaseIter<Outputs>),
+    table_key_images: &(impl DatabaseRo<KeyImages> + DatabaseIter<KeyImages>),
+    table_num_outputs: &(impl DatabaseRo<NumOutputs> + DatabaseIter<NumOutputs>),
+    table_rct_outputs: &(impl DatabaseRo<RctOutputs> + DatabaseIter<RctOutputs>),
+) -> Result<OutputOnChain, RuntimeError> {
     todo!()
 }
 
+//----------------------------------------------------------------------------------------------------
 /// TODO
 pub fn get_output_list() {
     todo!()
 }
 
+//---------------------------------------------------------------------------------------------------- `get_rct_num_outputs()`
 /// TODO
-pub fn get_rct_num_outputs() {
-    todo!()
+///
+/// # Example
+/// ```rust
+/// # use cuprate_database::{*, tables::*, ops::block::*, ops::output::*};
+/// // TODO
+/// ```
+#[doc = doc_error!()]
+#[inline]
+pub fn get_rct_num_outputs<'env, Ro, Rw, Env, Iter>(
+    env: &Env,
+    tx_ro: &Ro,
+) -> Result<u64, RuntimeError>
+where
+    Ro: TxRo<'env>,
+    Rw: TxRw<'env>,
+    Env: EnvInner<'env, Ro, Rw>,
+{
+    get_rct_num_outputs_inner(&env.open_db_ro::<RctOutputs>(tx_ro)?)
 }
 
+/// Internal function for [`get_rct_num_outputs_inner()`].
+#[inline]
+pub(super) fn get_rct_num_outputs_inner(
+    table_rct_outputs: &(impl DatabaseRo<RctOutputs> + DatabaseIter<RctOutputs>),
+) -> Result<u64, RuntimeError> {
+    // TODO: is this correct?
+    table_rct_outputs.len()
+}
+
+//---------------------------------------------------------------------------------------------------- `get_pre_rct_num_outputs()`
 /// TODO
-pub fn get_pre_rct_num_outputs() {
-    todo!()
+///
+/// # Example
+/// ```rust
+/// # use cuprate_database::{*, tables::*, ops::block::*, ops::output::*};
+/// // TODO
+/// ```
+#[doc = doc_error!()]
+#[inline]
+pub fn get_pre_rct_num_outputs<'env, Ro, Rw, Env, Iter>(
+    env: &Env,
+    tx_ro: &Ro,
+) -> Result<u64, RuntimeError>
+where
+    Ro: TxRo<'env>,
+    Rw: TxRw<'env>,
+    Env: EnvInner<'env, Ro, Rw>,
+{
+    get_pre_rct_num_outputs_inner(&env.open_db_ro::<Outputs>(tx_ro)?)
+}
+
+#[doc = doc_inner!(get_pre_rct_num_outputs)]
+#[inline]
+pub fn get_pre_rct_num_outputs_inner(
+    table_outputs: &(impl DatabaseRo<Outputs> + DatabaseIter<Outputs>),
+) -> Result<u64, RuntimeError> {
+    // TODO: is this correct?
+    table_outputs.len()
 }
```

Comment:
@Boog900 @SyntheticBird45 

Here's the current pattern for functions in `ops/` - each function comes in 3 flavors:
- `fn()` -> inputs/outputs 1 thing (opens tables for caller)
- `fn_bulk()` -> inputs/outputs N things (opens tables for caller)
- `fn_inner()` -> is the actual implementation (takes in already opened tables)

For example: `get_block()`, `get_block_bulk()`, `get_block_inner()`.

Some open questions:
1. Should `fn()` just be replaced with `_inner()` - as in, should we make the _caller_ open the tables and pass them instead? It would rid the need of an `_inner()`
2. Should these `_inner()` functions be public? Here's an example of what it would look like:

![image](https://github.com/Cuprate/cuprate/assets/101352116/9164964e-122b-4692-b0f0-165945828a80)

Reply:
IMO I think the answer to 1 should be yes - we should take the already opened tables instead of `EnvInner + Tx`.

- The function only takes what it needs
- More flexible for callers
- No lifetime + generics on all of our functions

Only tradeoff I think is there is there's more input set-up for the caller where many tables are passed - thankfully our `impl Tables` saves us from naming all tables:
```rust
//--------------------------------------- Before
pub fn get_block<'env, Ro, Rw, Env>(
    env: &Env,
    tx_ro: &Ro,
    height: BlockHeight,
) -> Result<VerifiedBlockInformation, RuntimeError>
where
    Ro: TxRo<'env>,
    Rw: TxRw<'env>,
    Env: EnvInner<'env, Ro, Rw>;

//--------------------------------------- After
pub fn get_block(
    tables: &impl Tables,
    height: BlockHeight,
) -> Result<VerifiedBlockInformation, RuntimeError>;
```

Reply:
2. No. It makes more sense that `fn_inner()` stays private. Even if its cheaper, then I don't see why we would use`fn()`. We could simply enforce opening everything required before  `fn()`. It would simply be moving the problem... So I public `fn()` needs to be defined, that internally use the `fn_inner()`

Reply:
Otherwise the functions flavors looks good and intuitive for me. Overall design is good :+1:

Reply:
To be clear I don't like the flavors - I want the caller to handle opening stuff so that there's only 1 minimal/flexible function.

We could remove the `_bulk()` functions as well since the caller is now free to `.collect()` into whatever container they want themselves since the function is cheap to call:
```rust
// Env.
let env = todo!();
let env_inner = env.env_inner()?;
let tx_ro = env_inner.tx_ro()?;

// Set-up the tables.
let tables: impl Tables = env_inner.open_tables(&tx_ro)?;

// Pass table directly and collect into whatever.
// This iterator returns early on error.
let blocks: Result<Vec<Block>, Error> = (0..100_000)
    .map(|height| get_block(height, &tables))
    .collect();
```

Reply:
I don't understand. What do you call the caller ?

Reply:
I agree we should just make the caller open the tables.

> We could remove the _bulk() functions as well since the caller is now free to .collect() into whatever container they want themselves since the function is cheap to call

Seems like a good idea. 

BTW If we are going to use `rayon`s parallel iterators we will probably need to use a [ThreadLocal]( https://docs.rs/thread_local/latest/thread_local/struct.ThreadLocal.html) which holds the tables as txs are not sync 

> What do you call the caller

The caller would be the service functions